### PR TITLE
Issue #61

### LINKED ISSUE #61: Manage activity
Allow user to manage their content activity manually.

Editing isn't required, but deleting and adding should be added (timestamp for activity should be modifiable)

### PR DESCRIPTION
**Pull request by stephaje**
### Changes made

- Allow clicking on activity entries to bring up an editor
- Allow for timestamp updating and deletion of activity entries
- Allow for different sized modals
- Misc: add build artifacts to gitignore
- Misc: change activity text to be in proper title case
- Misc: fix small slog missing key warning

### CODE REVIEW

File: server/activity.go
Comment (IRHM) [inline server/activity.go]: Log and error messages here (two below this one as well) should say updating activity.

File: server/activity.go
Comment (IRHM) [inline server/activity.go]: Is there anything in place to ensure an authenticated user can only update their own activity?

File: server/activity.go
Comment (IRHM) [inline server/activity.go]: Same as above; is there anything in place to ensure an authenticated user can only delete their own activity (user_id column check)?

File: src/lib/Activity.svelte
Comment (IRHM) [inline src/lib/Activity.svelte:1]: How come all of the string casing has been changed? The CSS handles the casing with `text-transform: capitalize;`, so it's easier to just keep it lowercase?

**Review by IRHM — CHANGES_REQUESTED** [review]
Hey @stephaje, thanks a lot for working on this, I really appreciate it!

I didn't have enough time to fully look through and run the changes right now, but I read over the code on github and added some comments.

File: server/activity.go
Comment (stephaje) [reply server/activity.go]: Good point, I'm still getting my bearings. Will update.

File: src/lib/Activity.svelte
Comment (stephaje) [reply src/lib/Activity.svelte:1]: I don't agree that's the best way to handle it.

- The title and description (at least in this context) should be in title case. Note that `text-transform: capitalize;` is not title case and also would lead to capitalization changes in stylized titles (e.x. [UFOria](https://www.imdb.com/title/tt0081674/?ref_=kw_li_tt), [ronny/lilly](https://www.imdb.com/title/tt8956332/?ref_=ttep_ep5)) if present.
- Therefore, the user of a Modal should be providing the strings in the exact format they want displayed.
- Converting a sentence from lower case to title case via a function is a little annoying, but [not impossible](https://github.com/gouch/to-title-case/blob/master/to-title-case.js), but also suffers from the issue of not respecting stylized titles.

Given all of the above, it's just more convenient to put the strings in title case manually.

Comment (IRHM) [reply src/lib/Activity.svelte:1]: @stephaje Thanks for explaining, we don't display any stylized titles/text in activity text, but I see what you mean and it does look better.

I have another note (hope they are not annoying), but I'll add it directly on the code so it's less confusing in a moment.

File: src/lib/Activity.svelte
Comment (IRHM) [inline src/lib/Activity.svelte]: The type of `text` can also be `undefined`, it shouldn't, but to avoid any errors, we could update the type to `string | undefined` and add a check to the top of the function to return `"unknown"` if the text is undefined.

File: src/routes/(app)/movie/[id]/+page.svelte
Comment (IRHM) [inline src/routes/(app)/movie/[id]/+page.svelte:232]: Also, we should add the new wListId prop to the <Activity/> components in tv and game pages. Currently the activity client state is only updated for tv.

File: src/lib/ActivityEditor.svelte
Comment (IRHM) [inline src/lib/ActivityEditor.svelte]: We have a `.danger` class for buttons that we could use here instead to keep the styles uniform.

File: src/lib/ActivityEditor.svelte
Comment (IRHM) [inline src/lib/ActivityEditor.svelte]: If using `.danger` class, we could remove this.

File: src/vars.scss
Comment (IRHM) [inline src/vars.scss:1]: If using `.danger` class, we could remove the delete-color variable.

File: server/activity.go
Comment (IRHM) [inline server/activity.go:95]: We could add a check here to ensure `res.RowsAffected` is bigger than 0. This will ensure we return an error if we didn't actually update anything (return "activity not found" error or something like that).

Currently we return a 200 OK if we try updating activity we don't own or doesn't exist.

Same for deleteActivity.

File: server/routes.go
Comment (IRHM) [inline server/routes.go]: We might want to rename `context` to `c` to keep it uniform with the rest of the handler funcs.

Same for delete below.

File: server/routes.go
Comment (IRHM) [inline server/routes.go:520]: After the `err == nil` check, we need to return an error similar to we do above:

```
c.AbortWithStatusJSON(http.StatusBadRequest, ErrorResponse{Error: err.Error()})
```

File: server/routes.go
Comment (IRHM) [inline server/routes.go]: In places where we just return a 400 status (no error msg returned to user), it may be beneficial to add an error log.

If we ever got an unexpected 400 in the future, it would be a big help to see the error in the server log.

File: src/lib/Activity.svelte
Comment (IRHM) [inline src/lib/Activity.svelte:219]: We could add `cursor: pointer` to this button so it's more obvious it's clickable.

**Review by IRHM — CHANGES_REQUESTED** [review]
Thanks for looking at the comments I made before.

A lot of small comments, of course no rush to get them done, feel free to call me a poop head and I can help out with the changes if you'd like.

Apart from the comments, it's looking and working good so far!

**Comment by stephaje** [comment]
All your comments looks reasonable to me. I'll work on changes sometime in the next ~24h

File: server/routes.go
Comment (stephaje) [reply server/routes.go]: I don't agree on this one. I think the others should change. There is a very small benefit and a lot of drawbacks to using shortened variable names. Robert C. Martin has an entire chapter on naming in [Clean Code](https://www.goodreads.com/book/show/3735293-clean-code) which I believe to be the best take on the subject.

I recognize this is counter to [go's own recommendations](https://go.dev/talks/2014/names.slide#6), I just don't agree with them.

I'll go either way on it, just let me know.

Comment (IRHM) [reply server/routes.go]: @stephaje I personally prefer keeping it short as long as its still quickly understandable (unfortunately dont have 30 buckaroos for Robertos book to see his points).

But regardless of recommendations, in this case, it's probably for the best to stick with the shorter variant `c` so it's inline with the other handlers, instead of refactoring the whole file (routes.go is a chubster).

**Review by IRHM — DISMISSED** [review]
Very happy with the changes, thanks again for doing this!

There's one more small bug I've just noticed. If I type an invalid date and click update, theres an error logged in devtools console (`Uncaught RangeError: invalid date at api.ts:174:23`), and the Updating notification stays there indefinitely.

image

I'll let you decide if you'd like to add something to handle that, I'm happy to merge as is if not and add something a bit later.

_(I think it should let you merge this after i approve)_

**Comment by stephaje** [comment]
Yeah, I'll take a look at it later this week. I borked my dev environment so I need some time to get things back together.

**Comment by stephaje** [comment]
Hmm, I actually can't reproduce that. It just lets me set it without error.

image

If you want to write a follow up issue I'm fine with that, but I'd need more info to diagnose. I've added a little bit of client validation in the meantime, which can be bypassed, but if people really want to break the application like that, I say let them.

**Review by IRHM — APPROVED** [review]
@stephaje I agree with you. Funny it lets you set the year 20, I can't explain why my browser breaks.

Anyways good work! All looks great, I'm happy to get it merged.

**Pull request merged**
